src/message_bus.zig

```diff
@@ -73,60 +191,413 @@ pub const MessageBus = struct {
+        var copy = self.replicas[self.server.replica].send_queue;
+        self.replicas[self.server.replica].send_queue = .{};
+        while (copy.pop()) |envelope| {
+            self.server.on_message(envelope.message);
+            envelope.message.references -= 1;
+            self.message_bus.gc(envelope.message);
+            self.message_bus.allocator.destroy(envelope);
+        }
     }
 
     pub fn create_message(self: *MessageBus, size: u32) !*Message {
         assert(size >= @sizeOf(Header));
 
         var buffer = try self.allocator.allocAdvanced(u8, vr.sector_size, size, .exact);
         errdefer self.allocator.free(buffer);
-        std.mem.set(u8, buffer, 0);
+        mem.set(u8, buffer, 0);
 
         var message = try self.allocator.create(Message);
         errdefer self.allocator.destroy(message);
 
         self.allocated += 1;
 
         message.* = .{
-            .header = std.mem.bytesAsValue(Header, buffer[0..@sizeOf(Header)]),
+            .header = mem.bytesAsValue(Header, buffer[0..@sizeOf(Header)]),
             .buffer = buffer,
             .references = 0,
         };
 
         return message;
     }
 
-    fn destroy_message(self: *MessageBus, message: *Message) void {
-        assert(message.references == 0);
-        self.allocator.free(message.buffer);
-        self.allocator.destroy(message);
-        self.allocated -= 1;
-    }
+    /// Connection used to send data to another replica
+    /// TODO: This connection will be automatically re-created if it drops,
+    /// using exponential backoff and full jitter.
+    const ReplicaConnection = struct {
+        message_bus: *MessageBus,
+        address: std.net.Address,
+        /// A value of -1 indicates that the connection is not currently active
+        fd: os.socket_t = -1,
+        completion: IO.Completion = undefined,
+        send_queue: FIFO(Envelope) = .{},
+        /// Number of bytes of the current message that have already been sent.
+        bytes_sent: usize = 0,
+
+        fn send_message(self: *ReplicaConnection, envelope: *Envelope) void {
+            const queue_was_empty = self.send_queue.out == null;
+            self.send_queue.push(envelope);
+            // If the queue was not empty, the message will be sent after the
+            // messages currently being sent.
+            if (queue_was_empty) self.send();
+        }
+
+        fn connect(self: *ReplicaConnection) void {
+            assert(self.fd == -1);
+            self.fd = try os.socket(self.address.any.family, os.SOCK_STREAM | os.SOCK_CLOEXEC, 0);
+            self.message_bus.io.connect(
+                *ReplicaConnection,
+                self,
+                complete_connect,
+                &self.completion,
+                self.fd,
+                &self.address.any,
+                self.address.getOsSockLen(),
+            );
+        }
+
+        fn complete_connect(self: *ReplicaConnection, completion: *IO.Completion, result: ConnectError!void) void {
```

Comment:
Let's rename all connection callbacks from e.g. `complete_connect` to `on_connect()` so that we're not using `complete_` and `completion` so much on the same line. Using `on_` is nice and short, we can then keep most of these function signatures on one line if under 100 columns, and it makes the syscall name pop.